src/create-job-form.tsx

```diff
@@ -252,6 +254,9 @@ export function CreateJobForm(props: CreateJobFormProps): JSX.Element {
           formInput={formInput}
           fields={formFields}
         />
+        {props.customEnvironment &&
+          <props.customEnvironment state={state} setState={setState} />
```

Comment:
One thing that users would most likely need is to run their own custom validation on the extended form fields. This could be run on the `onBlur` event for the custom form fields and we would update the validity state accordingly and you could disable the `Run Job` button if the state for this is `false`. 

```tsx
// in your extension
const [isFormValid, setIsFromValid] = useState(false); // default value of false

<props.customEnvironment state={state} setState={setState} isFormValid={isFormValid} updateValidity={setIsFormValid} />

// ---------------------------------------------------------------------------------------------------------

// in our custom env form (basic example)
const handleBlur = (e) => {
  const value = e.target.value;
  let isValid = false;

  if (value conforms to RegExp) { isValid = true; } 
  else { isValid = false; }

  updateValidity(isValid);
}
```

This would act as front-end validation for the UI piece.
Let me know if you have any questions or concerns regarding this ask!

Reply:
This is a good idea. However, race conditions are possible with a naive boolean implementation. Take the following sequence of events:

1. user selects from a dropdown, then network request is made to validate it
2. user moves to the next text field, typing in an invalid email address. `setValid(false)` gets invoked.
3. network request in 1) completes, and it returns true. `setValid(true)` gets invoked.

This sequence of events causes `valid` to be set to `true` despite the email field being invalid.

We need additional information as to which fields are causing the error, since a field's will not want to set the form state to valid if the even if its validate callback returns true. Thus, the `valid` state should be a dictionary mapping form keys to booleans. The form's valid state depends on whether all of the keys are `true`.

Reply:
On the SM UI side we are maintaining a dict of mappings of form fields to error messages, where an empty string means no error essentially. For example:
```
  const [formErrors, setFormErrors] = useState({
    field1: '',
    field2: '',
    field3: '',
  });
```

Would it make more sense to the check for this validity when the `Run Job` action is clicked on? This way, we can run the validation on one pass, instead of whenever a form field is 'focused' out of.

Reply:
Note that we're moving our UI components to MUI, which supports validation. See docs: https://mui.com/material-ui/react-text-field/#validation

Reply:
How would a consumer of this extension be able to perform their own custom validation, even if you are using Material UI?

Reply:
It looks like custom validation is possible by setting an MUI TextField's error property. Example here: https://stackoverflow.com/questions/69287291/text-field-validation-in-material-ui-using-error-property

Reply:
@awaisabir 

> Would it make more sense to the check for this validity when the Run Job action is clicked on? This way, we can run the validation on one pass, instead of whenever a form field is 'focused' out of.

Not really, because then what's the point of client-side validation when the user has to "click" the submit button to see the failure (same UX as server-side validation)? Seems silly. The whole benefit of client-side validation is that we can fail fast, i.e. on blur events.

If this is a heavily desired feature (which I strongly doubt), then we can perhaps add a configuration option for this somehow. But until then, please perform field validation on blur events for consistency.

> How would a consumer of this extension be able to perform their own custom validation, even if you are using Material UI?

If you think about it, errors are really just additional form state. My idea for how to handle this is that your extension component would accept two props: `errors` and `setErrors`. `errors` is mapping between form key and its corresponding error message (as you had suggested), and `setErrors` is self-explanatory. You can then use an effect to listen to the field state.

```ts
useEffect(() => {
  if (isInvalid(props.state.key)) {
    props.setErrors((errors) => ({ ...errors, key: "something is wrong" }))
  } 
}, [props.state.key])

return (
  ...
  <TextField value={...} onChange={...} {...(props.errors.key && { error: true, helperText: props.errors.key }) />
)
```

Perhaps there's other ways of doing this (since curiously it doesn't look like `TextField` exposes a `validate` or `onValidate` callback prop), but this was my first thought.

Reply:
@dlqqq 

> > Would it make more sense to the check for this validity when the Run Job action is clicked on? This way, we can run the validation on one pass, instead of whenever a form field is 'focused' out of.
> 
> Not really, because then what's the point of client-side validation when the user has to "click" the submit button to see the failure (same UX as server-side validation)? Seems silly. The whole benefit of client-side validation is that we can fail fast, i.e. on blur events.

We can handle it `onBlur`. Not a big deal here. Just a suggestion as an alternative approach to see if that would make life easier perhaps.

> If you think about it, errors are really just additional form state. My idea for how to handle this is that your extension component would accept two props: errors and setErrors. errors is mapping between form key and its corresponding error message (as you had suggested), and setErrors is self-explanatory. You can then use an effect to listen to the field state.

This is what I was suggesting to @jweill-aws as well. On the extension side you will have another state that handles the validity of the form itself - in the shape of an object with `[fieldName]: 'error message'` and pass the state and setter as props through the extension point. On our end we can append to this with the setter for our fields through the `onBlur` handlers for them 👍🏽